src/main/java/blf/core/writers/CsvWriter.java

```diff
@@ -90,29 +93,62 @@ protected void deleteState() throws Throwable {
+    protected void writeTable(String fileNameSuffix, String tableName) {
+        final String fileNameSuffixNullErrorMsg = "The suffix of the CSV file is null.";
+        final String tableNameNullErrorMsg = "The CSV table name is null.";
+
+        final String exportStartInfoMsg = String.format("Export of CSV table %s started.", tableName);
+        final String exportFinishInfoMsg = String.format("Export of CSV table %s finished.", tableName);
+        final String exportErrorMsg = String.format("An error occurred while exporting of CSV table %s.", tableName);
+
+        LOGGER.info(exportStartInfoMsg);
+
+        if (fileNameSuffix == null) {
+            this.exceptionHandler.handleException(fileNameSuffixNullErrorMsg, new NullPointerException());
+        }
+
+        if (tableName == null) {
+            this.exceptionHandler.handleException(tableNameNullErrorMsg, new NullPointerException());
+        }
+
+        final Path path = Paths.get(this.getOutputFolder().toString(), String.format("%s_%s.csv", tableName, fileNameSuffix));
 
         final Map<String, ArrayList<Object>> table = this.tables.get(tableName);
-        try (BufferedWriter writer = new BufferedWriter(new FileWriter(path.toFile()))) {
+
+        final File csvFile = path.toFile();
+
+        try (
+            final FileWriter csvFileWriter = new FileWriter(csvFile);
+            final BufferedWriter csvBufferedWriter = new BufferedWriter(csvFileWriter)
+        ) {
+
             final List<String> columns = this.columnNames.get(tableName);
 
             final String header = columns.stream().collect(Collectors.joining(this.delimiter));
-            writer.write(header);
-            writer.newLine();
+            final int rowCount = this.rowCounts.get(tableName);
 
-            for (int i = 0; i < this.rowCounts.get(tableName); i++) {
-                final int index = i;
+            csvBufferedWriter.write(header);
+            csvBufferedWriter.newLine();
+
+            for (int i = 0; i < rowCount; i++) {
+                final int rowIndex = i;
                 final String row = columns.stream()
-                    .map(column -> table.get(column).get(index))
+                    .map(column -> table.get(column).get(rowIndex))
                     .map(this::asString)
                     .collect(Collectors.joining(this.delimiter));
-                writer.write(row);
-                writer.newLine();
+
+                csvBufferedWriter.write(row);
+                csvBufferedWriter.newLine();
             }
+
+        } catch (Exception e) {
+            this.exceptionHandler.handleException(exportErrorMsg, e);
         } finally {
-            LOGGER.info(String.format("Export of CSV table %s finished.", tableName));
+            LOGGER.info(exportFinishInfoMsg);
+
+            this.tables.remove(tableName);
+            this.rowCounts.remove(tableName);
+            this.columnNames.remove(tableName);
```

Comment:
Same problem will probably arise here. In safe batching mode, writeState() and then writeTable() will be called multiple times, but state has to be preserved between them.